phantombuster/web-browser-extension
# feat: collect additional LinkedIn user info

## 📝 What is this PR

This PR aims to serve as a "proof of concept" to see if it's possible to easily retrieve additional LinkedIn user data directly from the extension.

## 🏭 How does it work

LinkedIn's data is retrieved using the LinkedIn's web API, specifically the `mysettings-api/settingsApiMiniProfile` endpoint.

The following happens when the extension is loaded:
- When the browser extensions detects a LinkedIn cookie field, it fetches the associated user info using the LinkedIn API
- Once the user info is retrieved on the background script, a "FETCH_COOKIE_USER_INFO_FULFILLED" message is dispatched to the frontend containing the formatted user info

When a cookie field of the frontend changes, the following happens:
- The frontend dispatches a `FETCH_COOKIE_USER_INFO` event (`{ cookieName: string, website: string, value: string }`)
- Message is intercepted by the content script and forwarded to the background script
- Background script checks if the cookie requested by the React app matches a cookie stored in the browser
- If that's the case, we retrieve all the cookies associated to the website requested by the React app (only LinkedIn for now)
- We then use those cookies to find the LinkedIn CSRF token
- When the CSRF token is found, a request to the  `mysettings-api/settingsApiMiniProfile` endpoint is made
- The response is then validated and forwarded to the content script through a `userInfo` message
- The content script then sends back the user info to the PhantomBuster's React app using `window.dispatchEvent`
- The React app can then intercept this event using `window.addEventListener("PB_EXTENSION_MESSAGE", callback)`


## 🗒 Notes

This new feature introduces the use of LinkedIn's API. This can be potentially "unsafe" as it could be detected by LinkedIn if they start doing extensive security checks. It might be a good idea to introduce this feature to a subset of users at first (or only internally).

## 🧪 QA Process

- Build the extension locally & install it in your browser
- Open the Chrome extensions panel: chrome://extensions/
- Find the PhantomBuster DEV extension and click to open the "service worker"
- Go to a LinkedIn's Phantom setup on your local front environment and click "connect to linkedin"
- Make sure cookies are filled correctly like before
- Check the "service worker" window. The console should display a log with `Received LinkedIn user info`

You can also edit your frontend app locally and add the following code in FormInput/Text.tsx:
```typescript
	useEffect(() => {
		const handleExtensionMessage = (event: CustomEvent) => {
			if (event.type === "PB_EXTENSION_MESSAGE") {
				console.log("RECEIVED", event.detail)
			}
		}

		window.addEventListener("PB_EXTENSION_MESSAGE", handleExtensionMessage as EventListener)
		return () => window.removeEventListener("PB_EXTENSION_MESSAGE", handleExtensionMessage as EventListener)
	}, [])
```

## 👌 To check

-   [ ] There's no regression
-   [ ] Coding guidelines are followed ([Contributing](https://github.com/phantombuster/engineering-docs/blob/master/contributing.md), [Testing](https://github.com/phantombuster/engineering-docs/blob/master/testing.md))
